plugins/fpc/fu-fpc-device.c

```diff
@@ -411,6 +458,141 @@ fu_fpc_device_setup(FuDevice *device, GError **error)
 	return TRUE;
 }
 
+static gboolean
+fu_fpc_device_write_ff2_payload(FuDevice *device,
+				FuFirmware *firmware,
+				FuProgress *progress,
+				FwupdInstallFlags flags,
+				GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+	const guint8 *data = NULL;
+	gsize data_length = 0;
+	gsize data_offset = FPC_FF2_HDR_OFFSET;
+	guint8 direction = FPC_FF2_BLOCK_DATA_DIR_OUT;
+	guint8 block_meta_hdr = 0;
+	guint16 payload_len = 0;
+	guint32 blocks_num = 0;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	data = g_bytes_get_data(fw, &data_length);
+	if (data == NULL || data_length == 0) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_INVALID_DATA, "invalid input data");
+		return FALSE;
+	}
+
+	if (!fu_memread_uint32_safe(data,
+				    data_length,
+				    FPC_FF2_META_BN_OFFSET,
+				    &blocks_num,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+
+	g_debug("block num: %u", blocks_num);
+	fu_progress_step_done(progress);
+	data_offset += FPC_FF2_BLK_OFFSET;
+
+	for (guint i = 0; i < blocks_num; i++) {
+		if (!fu_memread_uint8_safe(data, data_length, data_offset, &block_meta_hdr, error))
+			return FALSE;
+
+		if (block_meta_hdr != FPC_FF2_BLOCK_META_TYPE)
+			return FALSE;
+
+		if (!fu_memread_uint8_safe(data,
+					   data_length,
+					   data_offset + FPC_FF2_BLK_DIR_OFFSET,
+					   &direction,
+					   error))
+			return FALSE;
+
+		if (!fu_memread_uint16_safe(data,
+					    data_length,
+					    data_offset + FPC_FF2_BLK_PAYLOAD_LEN_OFFSET,
+					    &payload_len,
+					    G_LITTLE_ENDIAN,
+					    error))
+			return FALSE;
+
+		payload_len += FPC_FF2_BLK_PAYLOAD_HDR_LEN;
+		data_offset += FPC_FF2_BLK_SEC_LINK_OFFSET;
+
+		if (direction == FPC_FF2_BLOCK_DATA_DIR_OUT) {
+			guint32 remaining = payload_len;
+			guint32 transfer_offset = 0;
+			guint32 transfer_length = FPC_FF2_BLK_SEC_LINK_LEN;
+			while (remaining) {
+				if (data_offset + transfer_offset + transfer_length >=
+				    data_length) {
+					g_set_error(error,
+						    FWUPD_ERROR,
+						    FWUPD_ERROR_WRITE,
+						    "wrong offeset to write");
+					return FALSE;
+				}
+				if (!fu_fpc_device_dfu_cmd(
+					self,
+					FPC_CMD_DFU_DNLOAD_FF2,
+					0,
+					(guint8 *)&data[data_offset + transfer_offset],
+					transfer_length,
```

Comment:
I'm a bit confused why the first chunk is 100 bytes in size, and then the next chunks are 4096 bytes in size..

Reply:
We combine commands and firmware image data in the data of new firmware. So we use the first chunk to get the whole data size that it would be sent including commands and exact image data.